homeassistant/components/fritzbox/__init__.py

```diff
@@ -1,90 +1,77 @@
-                                vol.Required(CONF_USERNAME): cv.string,
-                            }
-                        )
-                    ],
+        DOMAIN: vol.All(
+            cv.ensure_list,
+            [
+                vol.Schema(
+                    {
+                        vol.Required(CONF_HOST): cv.string,
+                        vol.Required(CONF_PASSWORD): cv.string,
+                        vol.Required(CONF_USERNAME): cv.string,
+                    }
                 )
-            }
+            ],
+            ensure_unique_hosts,
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the fritzbox component."""
-
-    fritz_list = []
-
-    configured_devices = config[DOMAIN].get(CONF_DEVICES)
-    for device in configured_devices:
-        host = device.get(CONF_HOST)
-        username = device.get(CONF_USERNAME)
-        password = device.get(CONF_PASSWORD)
-        fritzbox = Fritzhome(host=host, user=username, password=password)
-        try:
-            fritzbox.login()
-            _LOGGER.info("Connected to device %s", device)
-        except LoginError:
-            _LOGGER.warning("Login to Fritz!Box %s as %s failed", host, username)
-            continue
-
-        fritz_list.append(fritzbox)
+async def async_setup(hass, config):
+    """Set up the AVM Fritz!Box integration."""
+    if DOMAIN in config:
+        for entry_config in config[DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
 
-    if not fritz_list:
-        _LOGGER.info("No fritzboxes configured")
-        return False
+    return True
 
-    hass.data[DOMAIN] = fritz_list
 
-    def logout_fritzboxes(event):
-        """Close all connections to the fritzboxes."""
-        for fritz in fritz_list:
-            fritz.logout()
+async def async_setup_entry(hass, entry):
+    """Set up the AVM Fritz!Box platforms."""
+    fritz = Fritzhome(
+        host=entry.data[CONF_HOST],
+        user=entry.data[CONF_USERNAME],
+        password=entry.data[CONF_PASSWORD],
+    )
+    await hass.async_add_executor_job(fritz.login)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzboxes)
+    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}, CONF_DEVICES: set()})
+    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = fritz
 
     for domain in SUPPORTED_DOMAINS:
-        discovery.load_platform(hass, domain, DOMAIN, {}, config)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, domain)
+        )
 
     return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unloading the AVM Fritz!Box platforms."""
+    fritz = hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id]
+    return await hass.async_add_executor_job(fritz.logout)
```

Comment:
Does this return a boolean?

Eventually we also want to unload the platforms here.

Reply:
There is nothing to be unloaded within the platforms.

Reply:
The core will unload the platforms if we forward it here. Entities will be removed etc.